GitHub Action to compile `template.tex`

Dntfreitas opened the pull request:
This PR implements a GitHub action to compile the `template.tex` file and push the generated PDF file to the main branch. The action is run upon a push to the main branch or a pull request.

The push message contains the version defined in the file `NOVAthesisFiles/nt-version.sty`. Moreover, the `template.tex` is compiled using the `pdflatex` compiler.

joaomlourenco commented:
@Dntfreitas, thank you so much for this pull request.
I have not worked with github actions before.  Before I accept your pull request, please help me understand a couple of details:
- usually I compile with "latexmk -pdf -shell-escape" to be sure the PDF is ok.
- while the default status for the template is "working", but the example PDF is for "final".  Usually I run a script that patches the `1_novathesis.tex` with the right flags, makes the PDF and then restores the changed files.  Would you help me do that?

Finally, you're from University of Madeira.  Can we collaborate to add UMadeira to the list of supported schools? :)

Dntfreitas commented:
@joaomlourenco, please check that the action is now generating the "final" PDF. I changed the flags and the compiler to match what you said.

I am also adding a badge, associated with the status of the Action, in the `README.md` file. The badge will be similar to those that you already have in `README.md` file.

Dntfreitas commented:
@joaomlourenco, I have made all the changes. Please check if it's okay to merge.

Regarding this question:
> Finally, you're from University of Madeira. Can we collaborate to add UMadeira to the list of supported schools? :)
I cannot answer you right now, because I need an official answer from the University. I will get back to you as soon as possible.

joaomlourenco commented on .github/workflows/build-template.yml:
```diff
@@ -0,0 +1,45 @@
+      - name: Compile the LaTeX template
+        uses: xu-cheng/latex-action@v2
+        with:
+          compiler: latexmk
```
This is the default compiler.  And the flag in line 31 only works if the compiler is not changed.  I think the best is to remove this line.

joaomlourenco replied:
Also, the PDF I want as showcase is not the one in the default configuration.
I use the script below to generate the final PDF.   Can we use it as the "compiler" instead?  Would the action still work?

```bash
#!/usr/bin/env bash

MAKE=$(which gmake)
[[ -n "$MAKE" ]] || MAKE=make

CONF=Config/1_novathesis.tex

# save the config file
cp ${CONF} ${CONF}.OLD

# select options for showcase PDF: final phd document form FCT-NOVA, with final index (_índice remissivo_) and trimed book spine
perl -pi -e "s,.*\\\\ntsetup{doctype=.*},\\\\ntsetup{doctype=phd}," $CONF
perl -pi -e "s,.*\\\\ntsetup{school=.*},\\\\ntsetup{school=nova/fct}," $CONF
perl -pi -e "s,.*\\\\ntsetup{docstatus=.*},\\\\ntsetup{docstatus=final}," $CONF
perl -pi -e "s,.*\\\\ntsetup{spine=.*},\\\\ntsetup{spine=trim}," $CONF
perl -pi -e "s,.*\\\\ntsetup{print/index=.*},\\\\ntsetup{print/index=true}," $CONF

# build the PDF
$MAKE

# restore defaults
mv ${CONF}.OLD ${CONF}
```

Dntfreitas commented:
@joaomlourenco, after a looot o trial and error, I finally get it right (I think). I am using exactly the same script that you provided. 

Could you please compare the PDF files (the one in your repo and the one, generated by the Action, in my fork)?

Also, to confirm, in the script that you use to build the PDF, is the key `print/index` correct? Or should it be `printindex`?

joaomlourenco commented:
You are right!  My mistake!
The keys are changing. :)
In current version (6.x.y) it is (still) "printindex", but in version 7.0.0 (to be made public soon) it's gonna be `print/index`.  But we can do that change later when the v7.0.0 gets public.
May I ask you to fix that key in the script?
 
Also, I think we should avoid cluttering the root directory with working files. Do you know if the `build.sh` script can go into the `.github` directory?  Or it is going to mess up the github actions?  The alternative is to move it to a new hidden directory (e.g., `.Build/build.sh`)

Thank you for your suggestion and for the patience with all my “requirements”.  The PDF in your fork looks perfect and I'm looking forward to accept your pull request and have the PDF file generated automatically.  👍

Dntfreitas commented:
@joaomlourenco, I updated the script with the correct key (I.e., from the current template version).

Also, I changed the path of the script compilation for the hidden path suggested by you.

Let me know if we are missing something before the merge.

joaomlourenco commented:
@Dntfreitas it seems the compilation is failing.
May I ask you to help fixing the problem?   I added you to the collaborators (commiters) of the novathesis project., in case you want to test and fix it  _in pace_!  
Thanks!!

b-pereira commented:
this is great! I tried the same Github action, but I didn't manage to produce a PDF file. Does it produce an updated PDF file? Also, it is possible to add new targets to `make` to GitHub Actions? I was trying to add `latexdiff` to my workflow and generate a new PDF file with a timestamp. Any help with be appreciated. @joaomlourenco if this is (kinda is) off-topic feel free to move this to Q&A.

Dntfreitas commented:
@joaomlourenco, I was able to replicate the error by setting the _Required approvals_ ON. If you turn it OFF, the GH Action will work fine. 

Do we need the _Required approvals_ ON? If you define the collaborators carefully, I don't think we need it ON.

<img width="1119" alt="image" src="https://user-images.githubusercontent.com/8163918/221662330-8ea940ef-ec2d-4cc0-af3a-6928ac1d2cb1.png">

Dntfreitas commented:
Also, _Require a pull request before merging_ should be OFF. Could you please try it?

joaomlourenco commented:
@Dntfreitas 
I was able to (partially) solve the problem without changing the merging options, with a Private Access Token.  But another problem arose…
When the action commits the generated PDF into main, it reactivates the action and starts generating the PDF again, succeeds, commits, triggers the action again…. :(
Did it happen to you in the past?
Do you know how to deal with this?

Dntfreitas commented:
@joaomlourenco First of all, sorry, I didn't know that this action would cause us many problems... 

I saw that you solved the loop of commits. Just an FYI, this could also be done: https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs

joaomlourenco commented:
@Dntfreitas everything is working. Thank you!
